src/value.jl

```diff
@@ -1,145 +1,181 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
+# "categorical value" trait
+abstract type CatValueSupport end
+struct IsCatValue <: CatValueSupport end
+struct NotCatValue <: CatValueSupport end
+
+# by default types do not have "categorical value" trait
+iscatvalue(::Type) = NotCatValue
+iscatvalue(::Type{Union{}}) = NotCatValue # otherwise it dispatches to Type{<:...}
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# "categorical value" trait implementation for CategoricalValue and CategoricalString
+iscatvalue(::Type{<:CatValue}) = IsCatValue
+valtype(::Type) = Union{} # no value type if not a proper categorical value type
```

Comment:
Is this actually used?

Reply:
Yes, AFAIR it gets called for `categorical([null, null])` test.

Reply:
Is `valtype(Null)` called then? Or `valtype(Union{})`? Maybe we should better define another method to catch these cases before they are passed to `valtype`? Or just restrict this to `valtype(::Null)` or `valtype(::Union{})`? Defining this method for any type sounds weird.

Reply:
It turns out that I was wrong and in the current version we don't need this method, so I will ditch it for good.